### PR TITLE
pilot integration tests use all configured clusters

### PR DESCRIPTION
**Pull request by stevenctl**
New PR: https://github.com/istio/istio/pull/25923

Converts tests under tests/integration/pilot to utilize all clusters in the test environment. The target is currently hidden. 

[integ-pilot-multicluster-tests_istio/30](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/24549/integ-pilot-multicluster-tests_istio/30) took 18 mins and has all traffic tests + meshnetwork test converted. 

Note: each commit is well scoped - much easier to review commit-by-commit. Willing to breakout into smaller PRs but some portion of this will be needed as the "base"

The progress of converting tests, broken down by suite below. 

#### pilot
traffic:
- [X] TestReachability
- [X] TestMirroring
- [X] TestMirroringExternalService
- [X] TestSniffing
- [X] TestTrafficRouting
- [ ] TestTrafficShifting (simpler to make a multicluster specific test)
- [X] TestCrossClusterTrafficShifting

#### pilot_meshnetwork
- [X] TestAsymmetricMeshNetworkWithGatewayIP (VM discovers all ingress gateways, pods on all clusters discover VM) (need to confirm CI passed)

istioctl:
- [ ] TestTrafficShifting
- [ ] TestVersion
- [ ] TestDescribe
- [ ] TestWait
- [ ] TestAuthZCheck
- [ ] TestProxyConfig
- [ ] TestProxyStatus

Sidecar:
- [ ] TestAddToAndRemoveFromMesh
- [ ] TestServiceEntryDNS
- [ ] TestServiceEntryDNSNoSelfImport
- [ ] TestServiceEntryStatic
- [ ] TestSidecarListeners
- [ ] TestSidecarScopeIngressListener

#### pilot_revisions
- [ ] TestMultiRevision

#### pilot_locality
- [ ] TestDistribute
- [ ] TestFailover
- [ ] TestPrioritized

#### pilot_analysis
- [ ] TestAnalysisWritesStatus

#### pilot_ingress
- [ ] TestGateway
- [ ] TestIngress

	
#### pilot_cni
- [ ] TestCNIReachability

#### pilot_vm
- [ ] TestVmOS
- [ ] TestVmOSPost

### CODE REVIEW

**Comment by stevenctl** [comment]
Both mirroring tests and reachability tests passing with 3 clusters (under 20 mins):
https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/24549/integ-pilot-multicluster-tests_istio/13

**Comment by stevenctl** [comment]
https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/24549/integ-pilot-multicluster-tests_istio/19
Passes in ~25 mins with:
- Mirroring
- Reachability
- Sniffing
- Routing

**Comment by stevenctl** [comment]
Cross cluster traffic shifting is seeing issues:
```
        	* unexpected traffic weight for host server-0. Expected 34%, got 0% (thresold: 8.5%)
        	* unexpected traffic weight for host server-1. Expected 33%, got 0% (thresold: 8.5%)
        	* unexpected traffic weight for host server-2. Expected 33%, got 100% (thresold: 8.5%)
```

Trying to determine if this is because server-2 is an independent control-plane or if it's because its on a different network.

File: pkg/test/framework/components/echo/kube/deployment.go [hunk 1]
```diff
@@ -327,11 +334,11 @@ func generateYAMLWithSettings(cfg echo.Config, settings *image.Settings, cluster
...
 	var vmImage, istiodIP, istiodPort string
 	if cfg.DeployAsVM {
-		s, err := kube.NewSettingsFromCommandLine()
-		if err != nil {
-			return "", "", err
-		}
-		addr, err := istio.GetRemoteDiscoveryAddress("istio-system", cluster, s.Minikube)
+		var addr net.TCPAddr
+		err = retry.UntilSuccess(func() error {
+			addr, err = istio.GetRemoteDiscoveryAddress(ctx, "istio-system", cluster)
```
Comment (nmittler) [inline pkg/test/framework/components/echo/kube/deployment.go:339]: Use the system namespace from the configuration

File: pkg/test/framework/components/echo/kube/deployment.go [hunk 2]
```diff
@@ -16,15 +16,21 @@ package kube
...
 	"fmt"
+	"net"
 	"strconv"
 	"text/template"
+	"time"
 
 	"github.com/Masterminds/sprig"
 
+	"istio.io/istio/pkg/test/framework/resource"
+
```
Comment (nmittler) [inline pkg/test/framework/components/echo/kube/deployment.go:27]: nit: remove additional spacing

File: pkg/test/framework/components/environment/kube/flags.go [hunk 3]
```diff
@@ -201,7 +201,10 @@ func init() {
 	flag.StringVar(&kubeConfigs, "istio.test.kube.config", strings.Join(settingsFromCommandLine.KubeConfig, ":"),
 		"A comma-separated list of paths to kube config files for cluster environments (default is current kube context)")
 	flag.BoolVar(&settingsFromCommandLine.Minikube, "istio.test.kube.minikube", settingsFromCommandLine.Minikube,
-		"Indicates that the target environment is Minikube. Used by Ingress component to obtain the right IP address..")
+		"Indicates that the target environment is Minikube. Used by to obtain the right IP address for ingress gateway.")
```
Comment (nmittler) [inline pkg/test/framework/components/environment/kube/flags.go:204]: I assume that when running with kind, both will be true? AFAIK the `minikube` flag probably should have been renamed `noexternalloadbalancer` (or something similar)... it was never really about minikube, itself. Would it make more sense to just run with minikube=false when our environment uses metallb?

File: pkg/test/framework/components/environment/kube/settings.go [hunk 4]
```diff
@@ -80,6 +83,12 @@ func (s *Settings) GetControlPlaneClusters() map[resource.ClusterIndex]bool {
 	return out
 }
 
+// SupportsExternalIP returns whether or not Services will acquire a public ip. If not, NodePort and host IP must be
+// used to workaround this.
+func (s *Settings) SupportsExternalIP() bool {
```
Comment (nmittler) [inline pkg/test/framework/components/environment/kube/settings.go:88]: Isn't the question more `SupportsExternalLB`?

File: tests/integration/pilot/istioctl_test.go [hunk 5]
```diff
@@ -94,7 +95,7 @@ spec:
     - destination: 
         host: reviews
 `)
-			istioCtl := istioctl.NewOrFail(ctx, ctx, istioctl.Config{})
+			istioCtl := istioctl.NewOrFail(ctx, ctx, istioctl.Config{Cluster: ctx.Environment().Clusters()[0]})
```
Comment (nmittler) [inline tests/integration/pilot/istioctl_test.go:98]: You're requiring that clusters be set explicitly now? I wonder if we should just make a utility for this since I suspect we'll need it often. Maybe `ctx.DefaultCluster()` (I'd like to eventually see the `Environment` abstraction go away and just be part of `resource.Context`)? That will also make it easy to identify all the places that still need to be updated later on.

File: tests/integration/pilot/mirror_test.go [hunk 6]
```diff
@@ -185,59 +175,86 @@ func runMirrorTest(options mirrorTestOptions) {
...
-			var instances [3]echo.Instance
-			echoboot.NewBuilderOrFail(options.t, ctx).
-				With(&instances[0], echoConfig(ns, "a")). // client
-				With(&instances[1], echoConfig(ns, "b")). // target
-				With(&instances[2], echoConfig(ns, "c")). // receives mirrored requests
-				BuildOrFail(options.t)
-
-			if options.fnInjectConfig != nil {
-				options.fnInjectConfig(ns, ctx, instances)
+			builder := echoboot.NewBuilderOrFail(options.t, ctx)
```
Comment (nmittler) [inline tests/integration/pilot/mirror_test.go:178]: Would it help if we separated out the cluster from the initial `echo.Config`? We could have a number of clusters associated with the `Builder` and then add helpers to make working with it easier:

```golang
result := builder.echoboot.NewBuilderOrFail(options.t, ctx).
  // Apply these configs to all clusters in the environment
  // (maybe this is the default?)
  WithClusters(ctx.Environment().Clusters()).
  With(echoConfig(ns, "a").
  With(echoConfig(ns, "b").
  With(echoConfig(ns, "c").
  BuildOrFail(options.t)

// Return all of the "a" services indexed by the cluster.
allA := result.LookupByName("a")

// Return a single "a" Instance in the given cluster.
a0 := result.GetServiceOrFail(ctx, "a", ctx.Environment().Clusters()[0])

// Other methods?
```

WDYT?

Comment (stevenctl) [reply tests/integration/pilot/mirror_test.go:178]: In a lot of scenarios that is great, but sometimes we want to test traffic to specific clusters. For that we have to make the service names unique to the cluster in question.

File: pkg/test/framework/components/environment/kube/flags.go (hunk 3 above)
Comment (stevenctl) [reply pkg/test/framework/components/environment/kube/flags.go:204]: Agreed. I'm pretty sure all references to the minikube flag are self-contained to this repo and its wiki. I'll probably keep the minikube flag for backwards-compat.

File: tests/integration/pilot/istioctl_test.go (hunk 5 above)
Comment (stevenctl) [reply tests/integration/pilot/istioctl_test.go:98]: It's not a requirement

File: tests/integration/pilot/mirror_test.go (hunk 6 above)
Comment (stevenctl) [reply tests/integration/pilot/mirror_test.go]: @nmittler Check out the changes to echo builder; let me know if you have a better idea than SetupFn. I used that to handle the unique naming and setting the correct pilot instance. 

On that note, echo doesn't really use the pilot instance we pass after removing the native environment. Theres a lot that could be cleaned up if we just remove that from echo. WDYT?

File: pkg/test/framework/components/echo/echo.go [hunk 7]
```diff
@@ -43,14 +44,20 @@ import (
 //     2. Build the source and destination Instances in separate groups and then
 //        call `source.WaitUntilCallable(destination)`.
 type Builder interface {
+	// WithClusters makes subsequent calls to With configure an instance with the same
+	// config to each of the passed clusters. Use methods on the Result returned
+	// by Build to get instances from all clusters.
+	WithClusters([]resource.Cluster) Builder
```
Comment (nmittler) [inline pkg/test/framework/components/echo/echo.go:50]: I wonder, would it be better to just replace `WithClusters` with:

```golang
WithPerCluster(cfg Config, clusters ...Cluster)
```

This would make it more clear which clusters are targeted for the configuration.

File: pkg/test/framework/components/istio/operator.go [hunk 8]
```diff
@@ -569,8 +569,14 @@ func configureDirectAPIServiceAccessForCluster(ctx resource.Context, env *kube.E
 	if err != nil {
 		return fmt.Errorf("failed creating remote secret for cluster %s: %v", cluster.Name(), err)
 	}
-	if err := ctx.Config(env.ControlPlaneClusters(cluster)...).ApplyYAML(cfg.SystemNamespace, secret); err != nil {
-		return fmt.Errorf("failed applying remote secret to clusters: %v", err)
+	// Copy this secret to all control plane clusters.
+	for _, remote := range env.ControlPlaneClusters() {
```
Comment (nmittler) [inline pkg/test/framework/components/istio/operator.go:573]: I think you can revert this change?

File: pkg/test/framework/components/istio/util.go [hunk 9]
```diff
@@ -69,16 +69,26 @@ func waitForValidationWebhook(ctx resource.Context, cluster resource.Cluster, cf
...
+func GetRemoteDiscoveryAddress(ctx resource.Context, namespace string, cluster resource.Cluster) (net.TCPAddr, error) {
+	env := ctx.Environment().(*kubeenv.Environment)
+	cpResCluster, err := env.GetControlPlaneCluster(cluster)
+	if err != nil {
+		scopes.Framework.Errorf("failed getting control-plane for cluster %d; trying to use it as a control-plane cluster", cluster.Index())
+		cpResCluster = cluster
+	}
+	cpCluster := cpResCluster.(kubeenv.Cluster)
+
+	// TODO(landow) deal with ingress circular dependency; only unique code here is getting the control-plane cluster
```
Comment (nmittler) [inline pkg/test/framework/components/istio/util.go:81]: I think we should move ingress and pilot into this package eventually ... WDYT?

File: pkg/test/framework/scope.go [hunk 10]
```diff
@@ -104,6 +102,16 @@ func (s *scope) get(ref interface{}) error {
 		}
 	}
 
+	if s.parent != nil {
```
Comment (nmittler) [inline pkg/test/framework/scope.go:105]: Is this a bug fix? If so, mind splitting it out to its own PR?

Comment (stevenctl) [reply pkg/test/framework/scope.go:105]: https://github.com/istio/istio/pull/25134

File: pkg/test/framework/components/istio/operator.go (hunk 8 above)
Comment (stevenctl) [reply pkg/test/framework/components/istio/operator.go:573]: Yeah, it was a big rebase.

File: pkg/test/framework/components/echo/echo.go (hunk 7 above)
Comment (stevenctl) [reply pkg/test/framework/components/echo/echo.go:50]: See https://github.com/istio/istio/pull/25139
